pandas/core/arrays/base.py

```diff
@@ -655,15 +657,65 @@ def copy(self, deep=False):
         raise AbstractMethodError(self)
 
     # ------------------------------------------------------------------------
-    # Block-related methods
+    # Printing
     # ------------------------------------------------------------------------
+    def __repr__(self):
+        from pandas.io.formats.printing import format_object_summary
+
+        template = (
+            u'{class_name}'
+            u'{data}\n'
+            u'Length: {length}, dtype: {dtype}'
+        )
+        # the short repr has no trailing newline, while the truncated
+        # repr does. So we include a newline in our template, and strip
+        # any trailing newlines from format_object_summary
+        data = format_object_summary(self, self._formatter(), name=False,
+                                     trailing_comma=False).rstrip()
+        class_name = u'<{}>\n'.format(self.__class__.__name__)
+        return template.format(class_name=class_name, data=data,
+                               length=len(self),
+                               dtype=self.dtype)
+
+    def _formatter(self, boxed=False):
+        # type: (bool) -> Callable[[Any], Optional[str]]
+        """Formatting function for scalar values.
+
+        This is used in the default '__repr__'. The returned formatting
+        function receives instances of your scalar type.
+
+        Parameters
+        ----------
+        boxed: bool, default False
+            An indicated for whether or not your array is being printed
+            within a Series, DataFrame, or Index (True), or just by
```

Comment:
Doesn't Index also use `False` like Array? (as I thought this was to satisfy the difference between Series/DataFrame vs Index)

Reply:
Index is hypothetical at this point, until we support EA-backed indexes. This isn't currently called by our index reprs. So we can choose to define it how we want.  Or we can make this a `box=None/Series/Index` argument.

Reply:
For external EAs, yes. But eg PeriodArray is stored in an Index, and there `box` is used for this purpose, no?

Reply:
It was not. Done in 3825aeb7f61b4d350888882689bce521b4a603ce

Reply:
I would add this arg to the Index formatters as well for compatiblity.

Reply:
I'm not sure I follow the need for adding it to the index formatters, since indexes are boxed by definition.